sys/net/gnrc/link_layer/lwmac/lwmac.c

```diff
@@ -0,0 +1,745 @@
+        break;
+
+    /* Receiving incoming data */
+    case RECEIVING:
+        rtt_handler(LWMAC_EVENT_RTT_PAUSE); /* No duty cycling while TXing */
+        _set_netdev_state(&lwmac, NETOPT_STATE_IDLE);  /* Power up netdev */
+        break;
+
+    case STOPPED:
+        _set_netdev_state(&lwmac, NETOPT_STATE_OFF);
+        break;
+
+    /*********************** Control states ***********************************/
+    case START:
+        rtt_handler(LWMAC_EVENT_RTT_START);
+        lwmac_set_state(LISTENING);
+        break;
+
+    case STOP:
+        rtt_handler(LWMAC_EVENT_RTT_STOP);
+        lwmac_set_state(STOPPED);
+        break;
+
+    case RESET:
+        LOG_WARNING("Reset not yet implemented\n");
+        lwmac_set_state(STOP);
+        lwmac_set_state(START);
+        break;
+
+    /**************************************************************************/
+    default:
+        LOG_DEBUG("No actions for entering state %u\n", newstate);
+        return;
+    }
+
+    lwmac_schedule_update();
+}
+
+/******************************************************************************/
+
+/* Main state machine. Call whenever something happens */
+// TODO: Don't use global variables
+bool lwmac_update(void)
+{
+    lwmac.needs_rescheduling = false;
+
+    switch(lwmac.state)
+    {
+    case SLEEPING:
+
+        /* If a packet is scheduled, no other (possible earlier) packet can be
+         * sent before the first one is handled, even no broadcast
+         */
+        if( !lwmac_timeout_is_running(&lwmac, TIMEOUT_WAIT_FOR_DEST_WAKEUP) ) {
+
+            /* Check if there are broadcasts to send and transmit immediately */
+            if(packet_queue_length(&(lwmac.tx.neighbours[0].queue)) > 0) {
+                lwmac.tx.current_neighbour = &(lwmac.tx.neighbours[0]);
+                lwmac_set_state(TRANSMITTING);
+                break;
+            }
+
+            lwmac_tx_neighbour_t* neighbour = _next_tx_neighbour(&lwmac);
+
+            if(neighbour != NULL) {
+                /* Offset in microseconds when the earliest (phase) destination
+                 * node wakes up that we have packets for. */
+                int time_until_tx = RTT_TICKS_TO_US(_ticks_until_phase(neighbour->phase));
+
+                /* If there's not enough time to prepare a WR to catch the phase
+                 * postpone to next interval */
+                if (time_until_tx < LWMAC_WR_PREPARATION_US) {
+                    time_until_tx += LWMAC_WAKEUP_INTERVAL_US;
+                }
+
+                time_until_tx -= LWMAC_WR_PREPARATION_US;
+                lwmac_set_timeout(&lwmac, TIMEOUT_WAIT_FOR_DEST_WAKEUP, time_until_tx);
+
+                /* Register neighbour to be the next */
+                lwmac.tx.current_neighbour = neighbour;
+
+                /* Stop dutycycling, we're preparing to send. This prevents the
+                 * timeout arriving late, so that the destination phase would
+                 * be missed. */
```

Comment:
> I guess the OS will check the related sleeping flag in another thread and stop the CPU when it find the flag is set up, right?  

Yes, this is exactly the case. Have a look at the [idle thread](https://github.com/RIOT-OS/RIOT/blob/9a0f3a9e404e12df6d02d317e545eebe8d7b9aa5/core/kernel_init.c#L64-L80). To activate low power sleeping in RIOT, you have to adapt the 'else' branch. The function of `lpm_prevent_sleep` should also be clear now. Only after all bits are cleared, RIOT attempts to stop the cpu.


> 2) the second question is somehow related to the first question. Assume in a "no traffic" scene in LwMAC, LwMAC will only wake up (```lpm_prevent_sleep |= LWMAC_LPM_MASK```) in the wakeup period, and sleep (```lpm_prevent_sleep &= ~(LWMAC_LPM_MASK)```) in the sleep period. So the CPU and along with all the threads will only run in the wake up period (WP) and stops in the sleep period, right? like the following

This depends on the other threads. Imagine there's another thread (or driver or whatever) that sets it's bit in `lpm_prevent_sleep`, then RIOT will not stop the cpu until all bits are cleared. You can see this as a request to RIOT to "please not stop the cpu until I'm done". If there's noone requesting such a wake lock, then it will be like you have described it.


Just to make it clear, I'm not sure if you have understood it correctly: `lpm_prevent_sleep |= LWMAC_LPM_MASK` does *not* wakeup the cpu, but requests to not stop the cpu when the idle thread is scheduled the next time. The wakeup is performed by an RTT alarm whose handler then sets `lpm_prevent_sleep |= LWMAC_LPM_MASK`.